### PR TITLE
Add e2e test for `aten.log_softmax_back_data` op

### PR DESCRIPTION
**Pull request by pashu123**
`aten.log_softmax_back_data` op lowering and required
tests has been added.

Signed-off-by: Prashant Kumar <prashant@nod-labs.com>

### CODE REVIEW

File: e2e_testing/torchscript/backprop.py [hunk 1]
```diff
@@ -50,3 +50,23 @@ def forward(self, out_grad, output):
...
+
+    @export
+    @annotate_args([
+        None,
+        ([-1, -1, -1], torch.float32, True),
+        ([-1, -1, -1], torch.float32, True),
+    ])
+    def forward(self, grad_output, output):
+        return torch.ops.aten._log_softmax_backward_data(grad_output,
+                                                     output,
```
Comment (ramiro050) [inline e2e_testing/torchscript/backprop.py:66]: nit: indentation is off here

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 2]
```diff
@@ -178,6 +178,51 @@ class DecomposeAten_SoftmaxBackwardDataOp
 };
 } // namespace
 
+// Aten_LogSoftmaxBackwardDataOp(gradOutput, output, dim) =>
+//    result = gradOutput - (exp(output) * sum(gradOutput, dim))
+namespace {
+class DecomposeAten_LogSoftmaxBackwardDataOp
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:184]: `LogSoftmaxBackward` and `SoftmaxBackward` seem to overlap quite a bit. This should be refactored so that both ops share the common parts. Remember to avoid copying and pasting code if possible.

**Review by ramiro050 — CHANGES_REQUESTED** [review]
Thanks for this change! A couple of comments below

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 3]
```diff
@@ -178,6 +178,51 @@ class DecomposeAten_SoftmaxBackwardDataOp
...
+
+    auto broadcastSizeType = Torch::ListType::get(Torch::IntType::get(context));
+    Value broadcastSize =
+        rewriter.create<AtenSizeOp>(loc, broadcastSizeType, output);
+    Value sumBroadcast =
+        rewriter.create<AtenBroadcastToOp>(loc, tensorType, sum, broadcastSize);
+    Value temp =
+        rewriter.create<AtenMulTensorOp>(loc, tensorType, expOut, sumBroadcast);
+
+    Value alpha =
```
Comment (cathyzhyi) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:215]: I think we now use AtenSubTensorOp often enough it probably worth to add a helper `createAtenTensorSub`  takes 2 tensors as arguments.

**Review by cathyzhyi — APPROVED** [review]
Looks good! Can you help to add a helper for tensors subtraction?

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 3 above)
Comment (pashu123) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:215]: Cool, I will add that.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 2 above)
Comment (pashu123) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:184]: Yes, thanks.

File: e2e_testing/torchscript/backprop.py (hunk 1 above)
Comment (pashu123) [reply e2e_testing/torchscript/backprop.py:66]: which Python-formatter do we use?

Comment (ramiro050) [reply e2e_testing/torchscript/backprop.py:66]: I'm not sure if there is a python formatter like clang-format. However, the style on these test files has usually been to align any list of arguments on the same column. Depending on your editor, this should happen if you hit tab on each argument line

Comment (cathyzhyi) [reply e2e_testing/torchscript/backprop.py:66]: I've been using  `yapf --in-place --recursive --style="{indent_width: 4}"`. it forces the pep8 format.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 3 above)
Comment (pashu123) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: Done.

Comment (ramiro050) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp]: This one can also use the helper :)

https://github.com/llvm/torch-mlir/blob/ea7a30f9b9b14f469024db35ce16bea2a5ee9548/lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp#L214-L217

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 2 above)
Comment (ramiro050) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:227]: I don't think this was fully addressed. In my opinion, the number of lines of code shared between `LogSoftmaxBackward` and `SoftmaxBackward` justifies a helper function. You should be able to create a helper function `decomposeSoftmaxBackwardDataLikeOp(Operation *op, PatternRewriter &rewriter)` that performs `return x - y * sum(x, dim)`, (assigning `x` and `y` accordingly depending on the `op`) so both `LogSoftmaxBackward` and `SoftmaxBackward` can share almost entirely all the code. This can make debugging a lot easier if a bug is found in the future, as well as making it easier to add features if needed.

Comment (pashu123) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:227]: I am really sorry for this. I will make sure that this gets merged. I am spinning up a PR.

Comment (pashu123) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:227]: @ramiro050  PR here: https://github.com/llvm/torch-mlir/pull/428
